src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs

```diff
@@ -259,5 +262,79 @@ private static CompletionItemRules MakeRule(bool importDirective, bool preselect
+
+        protected override CompletionItem CreateItem(
+            CompletionContext completionContext,
+            string displayText,
+            string displayTextSuffix,
+            string insertionText,
+            List<ISymbol> symbols,
+            SyntaxContext context,
+            bool preselect,
+            SupportedPlatformData supportedPlatformData)
+        {
+            if (IsAutoAddParenthesisBySemicolonEnabled(completionContext.Document))
+            {
+                var rules = CreateCompletionItemRules(completionContext, symbols, context, preselect);
+                var isInferredTypeDelegate = context.InferredTypes.Any(type => type.IsDelegateType());
+                var isObjectCreationTypeContext = context switch
+                {
+                    CSharpSyntaxContext csharpSyntaxContext => csharpSyntaxContext.IsObjectCreationTypeContext,
+                    _ => false
+                };
+
+                var shouldProvideParenthesisCompletion = symbols.All(symbol => symbol switch
+                {
+                    IMethodSymbol => !isInferredTypeDelegate,
+                    ITypeSymbol => isObjectCreationTypeContext,
+                    IAliasSymbol => isObjectCreationTypeContext,
+                    _ => false
+                });
+                var properties = ImmutableDictionary<string, string>.Empty
+                    .Add("ProvideParenthesisCompletion", shouldProvideParenthesisCompletion.ToString());
+
+                return SymbolCompletionItem.CreateWithNameKindAndId(
+                    displayText: displayText,
+                    displayTextSuffix: displayTextSuffix,
+                    symbols: symbols,
+                    // Always preselect
+                    rules: rules,
+                    contextPosition: context.Position,
+                    insertionText: insertionText,
+                    filterText: GetFilterText(symbols[0], displayText, context),
+                    supportedPlatforms: supportedPlatformData,
+                    properties: properties);
+            }
+            else
+            {
+                return base.CreateItem(
+                    completionContext,
+                    displayText,
+                    displayTextSuffix,
+                    insertionText,
+                    symbols,
+                    context,
+                    preselect,
+                    supportedPlatformData);
+            }
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = null, CancellationToken cancellationToken = default)
+        {
+            var insertionText = SymbolCompletionItem.GetInsertionText(item);
+            if (commitKey == ';' && SymbolCompletionItem.GetProvideParenthesisCompletion(item))
+            {
+                var symbols = await SymbolCompletionItem.GetSymbolsAsync(item, document, cancellationToken).ConfigureAwait(false);
```

Comment:
We don't need to encode symbols for all items, just applicable ones (i.e. shouldProvideParenthesisCompletion == true), we can then use the result of `GetSymbolsAsync` to decide if we should apply Parenthesis Completion. This way, we should be able to avoid encoding symbols in many cases.

Reply:
Change this part to treat types and method differently. So mark this as resolved.